pkg/migration-tool/planner/block.go

```diff
@@ -25,32 +25,100 @@ type Block struct {
 	pbar                  *progressbar.ProgressBar
 	timeseries            []*prompb.TimeSeries
+	stores                []store
+	numStores             int
 	numBytesCompressed    int
 	numBytesUncompressed  int
 	pbarMux               *sync.Mutex
 	plan                  *Plan // We keep a copy of plan so that each block hsa the authority to update the stats of the planner.
 }
 
-// Fetch starts fetching the samples from remote read storage based on the matchers.
-func (b *Block) Fetch(context context.Context, client *utils.Client, mint, maxt int64, matchers []*labels.Matcher) (*prompb.QueryResult, error) {
-	timeRangeMinutesDelta := (maxt - mint) / time.Minute.Milliseconds()
-	readRequest, err := utils.CreatePrombQuery(mint, maxt, matchers)
-	if err != nil {
-		return nil, fmt.Errorf("create promb query: %w", err)
+type store struct {
+	// Represent the order of a store. This helps to main the ascending order of time
+	// which is important while pushing data to remote-write storage.
+	id         int
+	mint, maxt int64
+}
+
+// initStores initializes the stores.
+func (b *Block) initStores() {
+	var (
+		proceed   = b.mint
+		increment = (b.maxt - b.mint) / int64(b.numStores)
+	)
+	for storeIndex := 0; storeIndex < b.numStores; storeIndex++ {
+		b.stores[storeIndex] = store{
+			id:   storeIndex,
+			mint: proceed,
+			maxt: proceed + increment,
+		}
+		proceed += increment
+	}
+	// To ensure that we cover the entire range, which may get missed due
+	// to integer division in increment, we update the maxt of the last store
+	// to be the maxt of the block.
+	b.stores[b.numStores-1].maxt = b.maxt
+}
+
+// Fetch starts fetching the samples from remote read storage based on the matchers. It takes care of concurrent pulls as well.
+func (b *Block) Fetch(ctx context.Context, client *utils.Client, mint, maxt int64, matchers []*labels.Matcher) (containsTs bool, err error) {
+	var (
+		totalRequests         = b.numStores
+		readRequests          = make([]*prompb.Query, totalRequests)
+		readResponses         = make([]*prompb.QueryResult, totalRequests)
+		contexts              = make([]context.Context, totalRequests)
+		cancelFuncs           = make([]context.CancelFunc, totalRequests)
+		timeRangeMinutesDelta = (maxt - mint) / time.Minute.Milliseconds()
+
+		doneChan = make(chan *utils.PrompbResponse, totalRequests)
+		errChan  = make(chan error, totalRequests)
+	)
+	for i := 0; i < totalRequests; i++ {
+		readRequest, err := utils.CreatePrombQuery(b.stores[i].mint, b.stores[i].maxt, matchers)
+		if err != nil {
+			return false, fmt.Errorf("create promb query: %w", err)
+		}
+		readRequests[i] = readRequest
+		cctx, cancelFunc := context.WithCancel(ctx)
+		contexts[i] = cctx
+		cancelFuncs[i] = cancelFunc
+		// Initiate concurrent fetching of data.
+		descriptionMsg := fmt.Sprintf("%s | time-range: %d mins", b.pbarDescriptionPrefix, timeRangeMinutesDelta)
+		go client.ReadChannels(contexts[i], readRequests[i], i, descriptionMsg, doneChan, errChan)
+	}
```

Comment:
You don't need the ReadRequests, ReadResponses or Context slices
```suggestion
		totalRequests         = b.numStores
		cancelFuncs           = make([]context.CancelFunc, totalRequests)
		timeRangeMinutesDelta = (maxt - mint) / time.Minute.Milliseconds()

		doneChan = make(chan *utils.PrompbResponse, totalRequests)
		errChan  = make(chan error, totalRequests)
	)
	for i := 0; i < totalRequests; i++ {
		readRequest, err := utils.CreatePrombQuery(b.stores[i].mint, b.stores[i].maxt, matchers)
		if err != nil {
			return false, fmt.Errorf("create promb query: %w", err)
		}
		cctx, cancelFunc := context.WithCancel(ctx)
		cancelFuncs[i] = cancelFunc
		// Initiate concurrent fetching of data.
		descriptionMsg := fmt.Sprintf("%s | time-range: %d mins", b.pbarDescriptionPrefix, timeRangeMinutesDelta)
		go client.ReadChannels(cctx, readRequests, i, descriptionMsg, doneChan, errChan)
	}
```

Reply:
Ah, thank you very much for this. This will save significant memory.